test/kad-dht.spec.js

```diff
@@ -686,54 +688,89 @@ describe('KadDHT', () => {
             connected[t] = true
             connect(others[i], others[t], cb)
           }, cb)
         }, cb),
+        // Make a connection from the origin node to each of the first 20 nodes
         (cb) => times(20, (i, cb) => {
           connected[i] = true
           connect(guy, others[i], cb)
         }, cb),
+        // Hash the key into the DHT's key format
         (cb) => kadUtils.convertBuffer(val, (err, rtval) => {
           expect(err).to.not.exist()
+
+          // Get the alpha (3) closest peers to the key from the origin's
+          // routing table
           const rtablePeers = guy.routingTable.closestPeers(rtval, c.ALPHA)
           expect(rtablePeers).to.have.length(3)
 
+          // Get all connected peers from the origin's peer book
           const netPeers = guy.peerBook.getAllArray().filter((p) => p.isConnected())
           expect(netPeers).to.have.length(20)
 
+          // The set of peers used to initiate the query (the closest alpha
+          // peers to the key that the origin knows about)
           const rtableSet = {}
           rtablePeers.forEach((p) => {
             rtableSet[p.toB58String()] = true
           })
 
+          // The ids of nodes that have connections
           const connectedIds = ids.slice(1).filter((id, i) => connected[i])
 
           series([
+            // Make the query
             (cb) => guy.getClosestPeers(val, cb),
+            // Find the closest connected peers to the key
             (cb) => kadUtils.sortClosestPeers(connectedIds, rtval, cb)
           ], (err, res) => {
             expect(err).to.not.exist()
+
+            // Query response
             const out = res[0]
+
+            // All connected peers in order of distance from key
             const actualClosest = res[1]
 
+            // Expect that the response includes nodes that are were not
+            // already in the origin's routing table (ie it went out to
+            // the network to find closer peers)
             expect(out.filter((p) => !rtableSet[p.toB58String()]))
               .to.not.be.empty()
 
+            // Expect that there were 20 peers found
             expect(out).to.have.length(20)
+
+            // The expected closest 20 peers to the key
             const exp = actualClosest.slice(0, 20)
 
+            // Expect the 20 peers found to be the 20 closest connected peers
+            // to the key
             kadUtils.sortClosestPeers(out, rtval, (err, got) => {
               expect(err).to.not.exist()
+
+              // TODO:
+              // With this change this expectation fails intermittently,
+              // because getClosestPeers() doesn't always find the closest 20
+              // peers (eg sometimes it will find 18 of the closest plus a
+              // couple of others that are not quite the closest).
+              // I'm not sure what the best approach to testing should be here?
               expect(countDiffPeers(exp, got)).to.eql(0)
```

Comment:
I'm not sure what the best approach to testing should be here?

Reply:
```js
// Make connections from each of the first 20 nodes to 16 of the
// remaining 80 nodes at random
```

Perhaps we should specify the topology for this so we can get deterministic results? If we pick nodes at random we're going to get random results. If we want a random dht network test, we can't have hard specifications because it's not a deterministic environment.  We'd have to do things like:

```js
// If results are greater than 20% different from the ideal set
// in a network of this size, there is problem and it needs to be fixed.
// If this test is failing DO NOT increase `.at.most` beyond 20% (4), there's
// a problem and it needs to be fixed.
expect(countDiffPeers(exp, got)).to.be.at.most(4)
```

I do think a better test is to specify the Topology. We should be able to get deterministic results that way correct? Is there any level of randomness that could affect that?

Reply:
That's true. An issue with this test is that it's not a very realistic topology. In reality, nodes close to a particular key should know about most of the other nodes close to the key. I will adjust the test to reflect that.

Reply:
We should do a deterministic test like jacob suggest, followed by a more real test, where we do not assume number like in this test.